### PR TITLE
feat: upload OCI image resources with skopeo

### LINKED ISSUE #1684: Allow Charmcraft to upload a rock directly from the file or directly between registries
### What needs to get done

Currently charmcraft requires an image digest on the local docker registry to upload an OCI image. When it does that, it extracts the OCI image to a tarball (equivalent to `docker image save`) and then uploads that.

Charmcraft should be able to take a rock file directly. For example:

```
$ charmcraft upload-resource my-charm my-resource ./my-rock_1.0.rock
```

This would do the equivalent of `skopeo copy oci-archive:my-rock_1.0.rock docker://registry.charmhub.io/my-rock:latest`

For copying between registries, Charmcraft should take something like:

```
$ charmcraft upload-resource my-charm my-resource docker://ghcr.io/canonical/charmed-postgresql:14.11-22.04_edge
```
This would do the equivalent of `skopeo copy docker://ghcr.io/canonical/charmed-postgresql:14.11-22.04_edge docker://registry.charmhub.io/my-rock:latest`

We have permission from @sergiusens to replace the current Python-based implementation with `skopeo`, but we must still be able to take:

```
$ charmcraft upload-resource my-charm my-resource  --filepath some-file
$ charmcraft upload-resource my-charm my-resource --image 31cf150b4523481202c1ff9b7b5d7f0b36729edad89d61242d8f1eb56b2912c0
```

cc @carlcsaposs-canonical 

### Why it needs to get done

Rockcraft outputs files directly

### PR DESCRIPTION
**Pull request by lengau**
Fixes #1684,  partial for #1685

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -66,12 +66,30 @@ jobs:
...
+          brew install skopeo
       - name: Configure environment
         run: |
           python -m pip install tox
           tox run --colored yes -m tests --notest
       - name: Run tests
+        shell: bash
         run: |
+          if [[ $(uname --kernel-name) == "Linux" ]]; then
+            # Ensure the version of skopeo comes from homebrew
```
Comment (tigarmo) [inline .github/workflows/tests.yaml:84]: how come skopeo needs to be from homebrew?

File: charmcraft/application/commands/store.py [hunk 2]
```diff
@@ -1936,87 +1937,95 @@ def run(self, parsed_args):
...
+                bases=bases,
+            )
         elif parsed_args.image:
+            emit.progress("Getting image")
+            emit.debug("Trying to get image from Docker")
+            image_service = self._services.image
+            if digest := image_service.get_maybe_id_from_docker(parsed_args.image):
+                emit.debug("Image is available via the local Docker daemon.")
+                source_path = f"docker-daemon:{digest}"
+            elif (image_path := pathlib.Path(parsed_args.image)).exists():
```
Comment (tigarmo) [inline charmcraft/application/commands/store.py:1956]: should you do this check first? It seems cheaper than checking docker

File: charmcraft/errors.py [hunk 3]
```diff
@@ -171,3 +175,22 @@ def __init__(self, extra_dependencies: Iterable[str]):
...
+class SubprocessError(CraftError):
+    """A craft-cli friendly subprocess error."""
+
+    @classmethod
+    def from_subprocess(cls, error: subprocess.CalledProcessError) -> Self:
+        """Convert a CalledProcessError to a craft-cli error."""
+        error_details = f"Full command: {shlex.join(error.cmd)}\nError text:\n"
+        if isinstance(error.stderr, str):
+            error_details += textwrap.indent(error.stderr, "  ")
+        else:
```
Comment (tigarmo) [inline charmcraft/errors.py:189]: `stderr` might also be `bytes` right?

**Review by tigarmo — APPROVED** [review]
nice!

File: charmcraft/errors.py (hunk 3 above)
Comment (lengau) [reply charmcraft/errors.py:189]: Yes, but the use case we have for this is always with `text=True`, so it shouldn't be an issue.

File: charmcraft/application/commands/store.py (hunk 2 above)
Comment (lengau) [reply charmcraft/application/commands/store.py:1956]: It is cheaper, but I don't want to break backwards compatibility here.

**Review by mr-cal — APPROVED** [review]
Looks good!

File: charmcraft/services/image.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        """Inspect an image with Skopeo and return the relevant metadata.
+
+        :param image: A skopeo-friendly image name
+        :returns: An OCIMetadata object containing metadata about the image.
+        """
+        try:
+            raw_manifest = self._skopeo.inspect(image, raw=True)
+            image_info = self._skopeo.inspect(image)
+        except errors.SubprocessError as exc:
+            raise errors.CraftError(
```
[the remark this answers is not in the document]
Comment (lengau) [reply charmcraft/services/image.py:135]: This is catching the error and replacing the error message with something more specific to the use case.

File: charmcraft/services/image.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    @override
+    def setup(self) -> None:
+        """Set up the image service."""
+        super().setup()
+        if not self._skopeo_path:
+            raise errors.CraftError(
+                "Could not find skopeo on the machine.",
+            )
+        logger.debug("skopeo path: %s", self._skopeo_path)
+        self._skopeo = utils.Skopeo(skopeo_path=self._skopeo_path)
```
[the remark this answers is not in the document]
Comment (lengau) [reply charmcraft/services/image.py:76]: Good catch - this is duplicated work.

File: charmcraft/application/commands/store.py (hunk 2 above)
Comment (tigarmo) [reply charmcraft/application/commands/store.py:1958]: makes sense! Should we add a note about backwards-compat here?

Comment (lengau) [reply charmcraft/application/commands/store.py:1958]: Done!

**Pull request merged**